File: DUUIWeb/src/lib/duui/component.ts
/**
 * This file contains constants, types, interfaces and functions related to DUUI components.
 */

import { v4 as uuidv4 } from 'uuid'

export const DUUIRemoteDriver = 'DUUIRemoteDriver'
export const DUUIDockerDriver = 'DUUIDockerDriver'
export const DUUISwarmDriver = 'DUUISwarmDriver'
export const DUUIUIMADriver = 'DUUIUIMADriver'
export const DUUIKubernetesDriver = 'DUUIKubernetesDriver'

export type DUUIDriver =
	| 'DUUIRemoteDriver'
	| 'DUUIDockerDriver'
	| 'DUUISwarmDriver'
	| 'DUUIUIMADriver'
	| 'DUUIKubernetesDriver'

export type DUUIDriverFilter = DUUIDriver | 'Any'
export const DUUIDriverFilters = [
	'Any',
	'DUUIRemoteDriver',
	'DUUIDockerDriver',
	'DUUISwarmDriver',
	'DUUIUIMADriver',
	'DUUIKubernetesDriver'
]

export const DUUIDrivers: string[] = [
	DUUIRemoteDriver,
	DUUIDockerDriver,
	DUUISwarmDriver,
	DUUIUIMADriver,
	DUUIKubernetesDriver
]

export type componentOptions =
	| 'use_GPU'
	| 'docker_image_fetching'
	| 'scale'
	| 'registry_auth'
	| 'constraints'
	| 'labels'
	| 'host'
	| 'ignore_200_error'

export interface DUUIComponent {
	oid: string
	id: string // Used for Drag & Drop
	name: string
	description: string
	tags: string[]
	driver: DUUIDriver
	target: string
	options: {
		use_GPU: boolean
		docker_image_fetching: boolean
		scale: number
		keep_alive: boolean
		registry_auth: {
			username: string
			password: string
		}
		constraints: string[]
		labels: string[]
		host: string
		ignore_200_error: boolean
	}
	parameters: any
	pipeline_id: string | null
	user_id: string | null
	index: number
}

/**
 * Create a blank component given a pipeline Id and an index in that pipeline.
 *
 * @param pipelineId The id of the pipeline the component is added to.
 * @param index The position in the pipeline.
 * @returns A default component.
 */
export const blankComponent = (pipelineId: string, index: number) =>
	<DUUIComponent>{
		oid: uuidv4(),
		id: uuidv4(),
		name: 'New Component ' + (index + 1),
		tags: [],
		description: '',
		driver: DUUIDockerDriver,
		target: '',
		options: {
			use_GPU: false,
			docker_image_fetching: true,
			scale: 1,
			keep_alive: false,
			registry_auth: {
				username: '',
				password: ''
			},
			constraints: [],
			labels: [],
			host: '',
			ignore_200_error: true
		},
		parameters: {},
		pipeline_id: pipelineId,
		user_id: null,
		index: index
	}

/**
 * Convert a DUUIComponent to a JSON object with some properties omitted.
 *
 * @param component the component to convert
 * @returns An object with relevant component properties.
 */
export const componentToJson = (component: DUUIComponent) => {
	return {
		name: component.name,
		tags: component.tags,
		description: component.description,
		driver: component.driver,
		target: component.target,
		options: component.options,
		parameters: component.parameters
	}
}

File: PythonClient/duui/objects/component.py
from __future__ import annotations
import json


class DUUIComponent:
    def __init__(self, data: dict) -> None:
        self.__data = {}

        for key, value in data.items():
            if key not in [
                "oid",
                "name",
                "description",
                "status",
                "categories",
                "created_at",
                "modified_at",
                "settings",
                "index",
                "pipeline_id",
                "user_id",
            ]:
                continue

            self.__data[key] = value

    def __str__(self) -> str:
        return json.dumps(self.__data, indent=2, ensure_ascii=False)

    def __repr__(self) -> str:
        return f"DUUIComponent(oid={self.oid}, modified_at={self.modified_at if self.modified_at > 0 else 'Never'})"

    @property
    def oid(self) -> str:
        return self.__data.get("oid", "")

    @property
    def name(self) -> str:
        return self.__data.get("name", "")

    @property
    def description(self) -> str:
        return self.__data.get("description", "")

    @property
    def categories(self) -> list[str]:
        return self.__data.get("categories", [])

    @property
    def status(self) -> str:
        return self.__data.get("status", "")

    @property
    def created_at(self) -> int:
        return self.__data.get("created_at", -1)

    @property
    def modified_at(self) -> int:
        return self.__data.get("modified_at", -1)

    @property
    def driver(self) -> str:
        return self.settings.get("driver", "")

    @property
    def target(self) -> str:
        return self.settings.get("target", "")

    @property
    def index(self) -> int:
        return self.__data.get("index", -1)

    @property
    def settings(self) -> dict:
        return self.__data.get("settings", {})

    @property
    def options(self) -> dict:
        return self.settings.get("options", {})

    @property
    def parameters(self) -> dict:
        return self.settings.get("parameters", {})

    @property
    def pipeline_id(self) -> str:
        return self.__data.get("pipeline_id", "")

    @property
    def user_id(self) -> str:
        return self.__data.get("user_id", "")
